### PR TITLE
Enable sharded checkpoint save and load (support local, sharded, and full state dicts for FSDP)

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?
Enables configuring what type of state dict to user for models and optimizers in `State`
We enable the user to set fsdp_config['state_dict_type'] to configure these types of state dicts. The types are:

* `'full'` (aka `torch.distributed.fsdp.StateDictType.FULL_STATE_DICT`) - the full unflattened, unsharded state dict materialized only on rank 0 with cpu offloading
* `'local'` (aka `torch.distributed.fsdp.StateDictType.LOCAL_STATE_DICT` - materializes just the local flattened shard of the state_dict on each rank
* `'sharded'` (aka `torch.distributed.fsdp.StateDictType.SHARDED_STATE_DICT`) - returns the unflattened, sharded state_dict.

when a user specifies `Trainer(model=my_model, optimiziers=my_optimizer, fsdp_config={..., 'state_dict_type': 'local'}, ...)`, then they enable local sharding for their model parameters and their optimizer states.

In order to enable this functionality this PR:
* Adds logic to `State.state_dict()` for model parameters and optim parameters to support the three FSDP state dict types
* logic for `State.load_state_dict()` for model and optim that supports state dict types
* modified `save_checkpoint` to allow non-zero ranks to save checkpoints
* modified `load_checkpoint` to allow non-zero ranks to load checkpoints 

Todo:
- [x] Add some documentation on how to use sharded checkpointing
- [x] unit tests

# Tests
did a bunch of manual tests 
* gpt-125m [save](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/1lgv1he0) [load](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/2245vy0d)
* gpt-30b [save](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/2cksdh6e)[ load](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/2vy3vi8i)
* gpt-70b [save](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/1xdcls10) [load](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/2vy3vi8i)

Also did manual tests to make sure proper error raised if:
* [ranks > checkpoint files](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/2f5zj9cn/overview)
* [ranks < checkpoint files](https://wandb.ai/mosaic-ml/evan-fsdp-ckpt-test/runs/1n2vsoce/overview)

# What issue(s) does this change relate to?

fix CO-1433
fix CO-1682

### CODE REVIEW

File: composer/core/state.py
Comment (bcui19) [inline composer/core/state.py:728]: Could you raise an error here if a user does pass in more than one optimizer? At some point we should nuke the > 1 optimizer code...

File: composer/core/state.py
Comment (bcui19) [inline composer/core/state.py:730]: A user could wrap their model as a FSDP model outside of composer and not pass in `fsdp_config`. This means the check would not pass. Might be good to rewrite the property to also check for `self.fsdp_config` in state rather than relying on this check?

Comment (bcui19) [reply composer/core/state.py:730]: It also feels a bit strange to check for FSDP being enabled via both `self.fsdp_config` and `self.fsdp_enabled`?

Comment (eracah) [reply composer/core/state.py:730]: I only do it to avoid a Pyright error!

Comment (eracah) [reply composer/core/state.py:730]: > A user could wrap their model as a FSDP model outside of composer and not pass in `fsdp_config`. This means the check would not pass. Might be good to rewrite the property to also check for `self.fsdp_config` in state rather than relying on this check?

good call will do that

File: composer/utils/checkpoint.py
Comment (dakinggg) [inline composer/utils/checkpoint.py:483]: update comment to reflect fsdp change

**Review by dakinggg — COMMENTED** [review]
Nice! Please add unit tests :) For manual test, could you also test a resumption run? i.e. i'd like to see that training for 100 batches is the same as training for 50 batches, checkpointing, and then resuming and training for another 50. Also would be good to check with all three types.

**Comment by eracah** [comment]
> Nice! Please add unit tests :) For manual test, could you also test a resumption run? i.e. i'd like to see that training for 100 batches is the same as training for 50 batches, checkpointing, and then resuming and training for another 50. Also would be good to check with all three types.

Already done: [see this](https://api.wandb.ai/links/mosaic-ml/8o9cwsdy)

**Comment by eracah** [comment]
> or manual test, could you also test a resumption run? i.e. i'd like to see that training for 100 batches is the same as training for 50 batches, checkpointing, and then resuming and training for another 50. Also would be good to check with all three types.

Already done [see this](https://api.wandb.ai/links/mosaic-ml/8o9cwsdy)

File: composer/trainer/dist_strategy.py
Comment (eracah) [reply composer/trainer/dist_strategy.py]: ok I updated `state.py` and `dist_strategy.py` to check and make sure `fsdp_config` is not None and `fsdp_config["state_dict_type"]` is set.

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: done

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: done

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: good question: torch forgot to put `ShardedStateDictConfig` in `torch/distributed/fsdp/__init__.py`, so I thought it didn't exist. Fixed.

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py:96]: done

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: fixed

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: corrigé

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: fixed in the State constructor

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: fixed in constructor: 
```python
self.fsdp_state_dict_type: Optional[str] = None
if self.fsdp_enabled:
    if self.fsdp_config is not None:
        self.fsdp_state_dict_type = self.fsdp_config.get('state_dict_type', 'full')
    else:
        self.fsdp_state_dict_type = 'full'
```

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: I guess I was using fsdp_state_dict_type for load and save. is it worth it having separate ones?

File: composer/core/state.py
Comment (eracah) [reply composer/core/state.py]: fixed

File: composer/utils/checkpoint.py
Comment (eracah) [reply composer/utils/checkpoint.py:483]: fixed

File: composer/utils/checkpoint.py
Comment (eracah) [reply composer/utils/checkpoint.py]: fixed

**Comment by eracah** [comment]
ok @dakinggg, I think I've addressed or asked for feedback on everything. Only things left are unit tests and docs. I will work on those now. Lmk if you have any follow-ups to my responses to your feedback

File: composer/core/state.py
Comment (dakinggg) [reply composer/core/state.py]: I guess I could imagine wanting to load in a sharded checkpoint and then write out a full checkpoint? but probably there is a better way to do that, and seems like that could be a follow on to this PR if needed.

**Review by dakinggg — COMMENTED** [review]
Looks good to me, thanks! Will wait for tests/docs for approval

**Comment by eracah** [comment]
Ok, @dakinggg, I added unit tests and docs. Give it another look when you get the chance.

**Review by dakinggg — APPROVED** [review]
LGTM

File: tests/trainer/test_sharded_checkpoint.py
Comment (eracah) [reply tests/trainer/test_sharded_checkpoint.py:104]: no I think that will just return rank 1's shard. This is because we set `rank0_only=True`. See this code I added to state.py: 
```python
if state_dict_type == 'full':
        state_dict_config = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
        fsdp_state_dict_type = StateDictType.FULL_STATE_DICT
```

Comment (dakinggg) [reply tests/trainer/test_sharded_checkpoint.py:104]: ah ok, in that case, maybe the test could confirm that?

File: docs/source/notes/distributed_training.rst
Comment (dakinggg) [reply docs/source/notes/distributed_training.rst:333]: maybe just "local sharded"?

Comment (eracah) [reply docs/source/notes/distributed_training.rst]: ok. fixed

File: tests/trainer/test_sharded_checkpoint.py
Comment (eracah) [reply tests/trainer/test_sharded_checkpoint.py:122]: ok added a test to check the rank 1 state dict

File: tests/trainer/test_sharded_checkpoint.py
Comment (eracah) [reply tests/trainer/test_sharded_checkpoint.py:234]: ok added a test to check for flattened vs unflattened in local vs sharded

**Comment by eracah** [comment]
@dakinggg, okay I added the tests you asked for. Can you give it one more quick once over and then I'll merge it?

File: tests/trainer/test_sharded_checkpoint.py
Comment (dakinggg) [inline tests/trainer/test_sharded_checkpoint.py:176]: this should be a skip, not an error, right? since it is in a test.

File: tests/trainer/test_sharded_checkpoint.py
Comment (dakinggg) [inline tests/trainer/test_sharded_checkpoint.py:225]: "total/" -> "total"

**Review by dakinggg — APPROVED** [review]
LGTM, thanks for adding the extra tests!

**Review by mvpatel2000 — APPROVED** [review]
LGTM

**Pull request merged**
